File: ruby/beer-song/beer_song.rb
module BeerSong
  class << self
    def recite(bottles, verses)
      @bottles = bottles
      result = ''
      loop do
        result << "#{pluralize.capitalize} on the wall, #{pluralize}.\n"
        result << "#{deal_with_beer} #{pluralize} on the wall.\n"
        verses -= 1
        break if verses < 1

        result << "\n"
      end
      result
    end

    def pluralize
      case @bottles
      when 0
        'no more bottles of beer'
      when 1
        '1 bottle of beer'
      else
        "#{@bottles} bottles of beer"
      end
    end

    def deal_with_beer
      if @bottles.zero?
        @bottles = 99
        'Go to the store and buy some more,'
      else
        @bottles -= 1
        "Take #{@bottles.zero? ? 'it' : 'one'} down and pass it around,"
      end
    end
  end
end

File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(number) do
    "#{count(number, true)} #{bottles(number)} on the wall, " <>
      "#{count(number)} #{bottles(number)}.\n" <>
      action_taken(number)
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range \\ 99..0)

  def lyrics(range) do
    range
    |> Enum.map(fn x -> verse(x) end)
    |> Enum.join("\n")
  end

  defp bottles(1), do: "bottle of beer"
  defp bottles(_), do: "bottles of beer"

  defp article(1), do: "it"
  defp article(_), do: "one"

  defp count(number, capitalize \\ false)
  defp count(0, true), do: "No more"
  defp count(0, false), do: "no more"
  defp count(number, _), do: number

  defp action_taken(0), do: "Go to the store and buy some more, " <> contemplate(99)
  defp action_taken(number), do: "Take #{article(number)} down and pass it around, " <> contemplate(number - 1)

  defp contemplate(number) do
    "#{count(number)} #{bottles(number)} on the wall.\n"
  end
end
